File: 07/day07.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	lines, err := readLines("input")

	if err != nil {
		os.Exit(1)
	}

	bags := makeBags(lines)
	fmt.Println(solve1(bags))
	fmt.Println(solve2(bags))
}

func solve1(bags map[string]map[string]int) int {
	n := 0
	seen := map[string]bool{}
	for k := range bags {
		if traverse(bags, k, seen) {
			n++
		}
	}
	return n
}

func solve2(bags map[string]map[string]int) int {
	seen := map[string]int{}
	return traverse2(bags, "shiny gold", seen)
}

func traverse(
	bags map[string]map[string]int,
	bag string,
	seen map[string]bool,
) bool {
	if v, ok := seen[bag]; ok {
		return v
	}

	for k := range bags[bag] {
		if k == "shiny gold" {
			seen[bag] = true
			return true
		}
	}

	for k := range bags[bag] {
		if traverse(bags, k, seen) {
			seen[bag] = true
			return true
		}
	}

	seen[bag] = false
	return false
}

func traverse2(
	bags map[string]map[string]int,
	bag string,
	seen map[string]int,
) int {
	if v, ok := seen[bag]; ok {
		return v
	}

	n := 0
	for k := range bags[bag] {
		v := bags[bag][k]
		n += v + (v * traverse2(bags, k, seen))
	}

	seen[bag] = n
	return n
}

func makeBags(lines []string) map[string]map[string]int {
	bags := map[string]map[string]int{}
	re := regexp.MustCompile(`(\d+)\s(\w+\s\w+)`)
	for _, l := range lines {
		split := strings.SplitN(l, " bags contain ", 2)
		match := re.FindAllStringSubmatch(split[1], -1)

		subBags := map[string]int{}

		if len(match) > 0 {
			for _, m := range match {
				subBags[m[2]] = atoi(m[1])
			}
		}

		bags[split[0]] = subBags
	}

	return bags
}

func readLines(path string) ([]string, error) {
	lines := []string{}

	file, err := os.Open(path)

	if err != nil {
		return lines, err
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines, scanner.Err()
}

func atoi(s string) int {
	n, err := strconv.Atoi(s)
	if err != nil {
		return 0
	}
	return n
}

File: 07/day07.py
#!/usr/bin/env python3

import itertools
import re

from functools import reduce


def main():
    with open("input", "r") as f:
        lines = f.read().splitlines()
    bags = dict(parse(i) for i in lines)
    print(solve1(bags))
    print(solve2(bags))


def solve1(bags):
    return sum(traverse(bags, k) for k in bags.keys())


def solve2(bags):
    return traverse2(bags, "shiny gold")


def traverse(bags, bag):
    if "shiny gold" in bags[bag].keys():
        return True

    for k, v in bags[bag].items():
        if traverse(bags, k):
            return True

    return False


def traverse2(bags, bag):
    return sum(v + (v * traverse2(bags, k)) for k, v in bags[bag].items())


def parse(s):
    main_colour, _sub_colours = s.split(" bags contain ", 2)
    reg = r"(\d+)\s(\w+\s\w+)"
    sub_colours = {col: int(n) for n, col in re.findall(reg, _sub_colours)}
    return main_colour, sub_colours


if __name__ == "__main__":
    main()
